service/history/query.go

```diff
@@ -113,6 +115,22 @@ func (q *queryImpl) completeQuery(
 	return nil
 }
 
+func (q *queryImpl) unblockQuery() error {
+	q.Lock()
+	defer q.Unlock()
+
+	if q.terminalState() {
+		return errAlreadyInTerminalState
```

Comment:
Should you just simply return nil here? Otherwise it does not seem there is a way out? Meaning that queryRegistryImpl.unblockQuery will not remove it from the map

Reply:
I think I prefer to return this error here because if you return nil it makes the logic in queryRegistry weird 

```
	if err := q.completeQuery(queryResult); err != nil {
		return err
	}
	delete(r.buffered, id)
	r.completed[id] = q
	return nil
```

If you return nil here then after ```complete``` is called "successfully" then you have to run checks for what state you are in. 

Also query does not does have to move to a terminal state to be removed from the map. The only way things get removed from queryRegistry is by having ```delete``` called. So the pattern for using query registry should be 

```
qr := getRegistry
queryID := qr.Buffer()
defer qr.Remove(queryID) // this removes from all maps regardless of the progress query has made
```

Reply:
Since ```delete``` force removes a query from map regardless of state I think the right pattern for coding state transitions is to return an error whenever the state transition did not occur.